### DISCUSSION TITLE
alarm-control-panel: Vacation Mode

### DISCUSSION THREAD

**Discussion by posixx** [home-assistant/architecture#500] [Entity Models] [answered]
## Context
Multiple alarmpanel vendors have a vacation mode apart from existing modes like away, night and home:

- Cytech Comfort (this one is used by me and tested though MQTT)
- ADT Security Systems
- DSC Alarm System
- ELK M1 Alarmsystem

We like to have this extra alarm status / mode added to the alarm-control-panel entity so users are able to do specific automation actions based on them being away for a longer time like close main water valves, specific light scenes or extra ventliation during specific hours

## Proposal

Adding the extra vacation mode to the alarm-control-panel entity as well as the MQTT alarm class. 
PR's related to this proposal: 
 - https://github.com/home-assistant/developers.home-assistant/pull/758
 - https://github.com/home-assistant/core/pull/45980 
 - https://github.com/home-assistant/frontend/pull/8326

and contains the following modifications:

**Services**
Add the following service:
```
alarm_arm_vacation:
  description: Send the alarm the command for arm vacation.
  fields:
    entity_id:
      description: Name of alarm control panel to arm vacation.
      example: "alarm_control_panel.downstairs"
    code:
      description: An optional code to arm vacation the alarm control panel with.
      example: "1234"
```
Along with a SUPPORT_ALARM_ARM_VACATION flag.

**States**
Add the state STATE_ALARM_ARMED_VACATION

**Device automations**
Add an arm_vacation device action
Add an is_armed_vacation device condition
Add an armed_vacation device trigger

**UI**
Add an "Arm Vacation" button to the alarm control panel card
Add an icon to represent the armed vacation state to the card and details dialog.

## Consequences

All existing alarmpanels that have integrations within HA will not be impacted by adding this new alarm state. The integration component of the alarmpanel determines if this state is used or not. 

Currently it is impossible to have automation actions based on vacation without some extra manual interaction. With this new state the alarmpanel this will be possible.

**Comment by posixx** [comment]
Ho does this work; when will this be discussed / follow'd up?

**Comment by MartinHjelmare** [comment]
There's unfortunately no formal process or timeline for that. There needs to be some interest from organization members to look at this and comment.

My initial thought is that 4/29 seems a bit low coverage for this feature to be added. Have you reviewed all existing integrations with alarm control panel platforms to gather the correct possible support level for this feature?

**Comment by posixx** [comment]
No i didn't. I don't know all integrations, i'm sure there will be more supporting vacation mode., i just searched google for some examples. 

@frenck suggested to open an architectural change request for this, but this change has no impact at all on all 29 alarm implementations other than making the new status available to panels that support it.

**Comment by MartinHjelmare** [comment]
Before adding a new feature to an entity integration we want to assess the support for this feature among both existing integrations and for other devices of the same device type in the wild.

It's easier to accept new features that we know are supported by a majority or at least large part of the integrations. So I'd advice to try and gather more data on this.

**Comment by posixx** [comment]
Apart from the list i provided i scanned the current alarm integrations and found the following to be supporting some sort of vacation mode:

- arlo
- envisalink
- homekit
- homematic
- honeywell
- lupus
- interlogix
- ness
- verisure
- wink
 
So 11 out of 29 intergrations are supporting vacation mode (didn't count in MQTT as that would be implemented with my pull request). Apart from that; home and/or night modes aren't supported either on a lot of panels which do not support vacation mode, so if these modes are supported by HA so should vacation mode in my opinion. The use cases for vacation mode are much higher than for instance, home mode.

**Comment by MartinHjelmare** [comment]
With 11/29 I'm ok with adding it. We should get some more opinions from other members before moving ahead.

**Comment by frenck** [comment]
For Verisure this is not a valid use-case. Verisure vacation/holiday mode is a setting for the alarm dispatch center to know who to contact between a certain period. It cannot be activated or read out from the alarm panel itself.

Nevertheless, I don't see much harm in adding such a mode.

**Comment by posixx** [comment]
According to this python module it should have some information available: https://github.com/persandstrom/python-verisure,

**Comment by frenck** [comment]
Yes, but it isn't an alarm panel state.

**Comment by posixx** [comment]
Ok clear!

**Comment by posixx** [comment]
Hello, any updates on implementing vacation mode?

**Comment by austinmroczek** [comment]
totalconnect API has no concept of vacation mode as described here.  There would be no impact to totalconnect by implementing this.

**Comment by OnFreund** [comment]
Risco doesn't have a formal vacation mode (AFAIK), but it has 6 different arming modes, and the user can set one of them to something that would make sense when going on vacation. The integration is planned to allow custom mapping of HA arming modes to Risco arming modes (https://github.com/home-assistant/core/pull/39218).
In short - this could make sense for the Risco integration.

**Comment by bdraco** [comment]
Seems fine to me.  ElkM1 supports this and I'd be happy to add it.

**Comment by posixx** [comment]
at what period a decision will be made for this? My pull request is becoming stale again,.

**Comment by posixx** [comment]
bunp..

**Comment by posixx** [comment]
As there seems no interest in this i've made a workaround and use a flag within my alarm to detect vacation mode.

**Comment by frenck** [comment]
I was looking throught the open PRs and came across <https://github.com/home-assistant/core/pull/38284>.

Yet, @posixx, this one is closed?

> As there seems no interest

I believe the verdict in this issue was rather approving.

3 Members (myself, @MartinHjelmare & @bdraco) approved of adding it and the reasoning for it with the overview of the integrations that could support it; was IMHO enough to go forward?

**Comment by MartinHjelmare** [comment]
Please also describe the frontend source changes needed for this feature. A frontend PR that implements this is required before we can merge the core PR.

**Comment by posixx** [comment]
There is no frontend change, just adding a alarm status. You can see all you need in the PR https://github.com/home-assistant/core/pull/38284

**Comment by MartinHjelmare** [comment]
As I understand the core PR, it's adding a new service and a new state. This should normally be exposed by the frontend.

It would help this architecture issue if we outline the proposed changes to the entity model here in the PR description. See eg this other issue:
https://github.com/home-assistant/architecture/issues/375

**Comment by posixx** [comment]
I'm sorry i don't understand what i need to add where..

**Comment by MartinHjelmare** [comment]
First, in this PR description, outline the new alarm control panel state and service and corresponding entity method that you propose to add.

Second, please describe how these new features will be exposed in the frontend. Normally, entity services are available to be run via the GUI. If we don't want this, please explain why so we can decide on this too.

**Comment by posixx** [comment]
@MartinHjelmare i tired to determine what you want me to add to this PR but i don't get it at all. The extra alarmstate is completely the same as, for instance, the nightmode implementation. There are no differences apart from the alarm state value. 

I would opt to use the nightmode implementations (core aswell as frontend) as reference for this new alarmstate.

**Comment by MartinHjelmare** [comment]
For arm away mode there's a button in the lovelace card to activate this mode. Should we add this for the vacation mode? This is the sort of discussion I'm after.

You haven't specifically described the APIs, and named them, that you want to add besides opening the core PR. I'd like you to do that here in the architecture issue top post so it's easier to understand and talk about this here.

Please describe the API in terms of service and state, which are core building blocks in Home Assistant.

**Comment by posixx** [comment]
Ok i updated the description of this PR.

I would opt for adding the button in lovelace to activate vacation mode. Also make visibility of this button customizable just as other states.

**Comment by OnFreund** [comment]
Sorry for interjecting, but I believe I understand what both sides here are after, so I'll try to tackle this:

**Services**
Add the following service:
```
alarm_arm_vacation:
  description: Send the alarm the command for arm vacation.
  fields:
    entity_id:
      description: Name of alarm control panel to arm vacation.
      example: "alarm_control_panel.downstairs"
    code:
      description: An optional code to arm vacation the alarm control panel with.
      example: "1234"
```

Along with a `SUPPORT_ALARM_ARM_VACATION` flag.

**States**
Add the state `STATE_ALARM_ARMED_VACATION`

**Device automations**
* Add an `arm_vacation` device action
* Add an `is_armed_vacation` device condition
* Add an `armed_vacation` device trigger

**UI**
* Add an "Arm Vacation" button to the alarm control panel card
* Add an icon to represent the armed vacation state to the card and details dialog.

Hope this helps.

**Comment by posixx** [comment]
Many thanks @OnFreund i updated the comments!

**Comment by posixx** [comment]
Ok when will this proposal going forward?

**Reply by frenck** [reply]
I think that is up to you, that said, if you agree on the outline set by @OnFreund in <https://github.com/home-assistant/architecture/discussions/500#discussioncomment-330413>.

**Reply by posixx** [reply]
what do you mean.. This was added to the context minutes after onfreund's post..

**Reply by frenck** [reply]
I mean, feel free to move forward on implementing this proposal in that case.

**Reply by posixx** [reply]
This is already implemented, see https://github.com/home-assistant/core/pull/38284. The code is there. The checks which didn't succeed i cannot fix as my knowldge of pyhton doesn't go that far

**Reply by frenck** [reply]
ok... Let's work from there more.
Additionally, the frontend part needs to be implemented and linked up with that PR.

**Reply by posixx** [reply]
Hmm, i don't have the knowledge to do that.. Try to find out when i have some time

**Reply by posixx** [reply]
Ok I looked at the frontend project, but i'm not going to fiddle into that. For someone familiar with this project it will take 2 minutes to add button and status for the alarmpanel card. Also for implementation these frontend changes aren't reuired; it tested without them and it worked without issues.

EDIT: as not every alamrpanel supports vacation mode i opt for leaving the fronend as-is. When alarm is put into vacationmode it shows on the frontend aswell as the armed icon (see my screenshots on the PR into). So good to go!

**Reply by frenck** [reply]
> as not every alamrpanel supports vacation mode i opt for leaving the fronend as-is. 

That is what the `SUPPORT_ALARM_ARM_VACATION` is for?
If an integration set supported feature flags, the frontend can react to that.

**Reply by posixx** [reply]
Yes i guess that will work.. Hopefully someone else can reach out and implement the frontend change!

**Reply by posixx** [reply]
Ok, Core aswell as Frontend changes are done. Please proceed and merge the PR's mentioned in the comments

**Reply by posixx** [reply]
Tell, me, why is a proposal which was don for 2 months now not merged?

**Reply by emontnemery** [reply]
The core PR needs to be rebased
Edit: Done

**Comment by posixx** [comment]
As both Core and Frontend PR's are completed this architectual change can now be merged into master

**Comment by posixx** [comment]
Will this be part of the new upcoming Beta release?

**Comment by posixx** [comment]
PR's are all completed. When will this merge into new beta release??

**Comment by frenck** [comment, marked as answer]
Backend & developer documentation has been merged, frontend is approved and awaiting merge.

**Reply by posixx** [reply]
Tnx frenk! Hopefully Bram can fix the build error on the frontend soon so it is complete.

**Reply by frenck** [reply]
Don't thank me :) I was @emontnemery that pushed it over the finish line 😉
